rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -1248,19 +1250,21 @@ impl Program {
 
     /// TODO: Allow this to return an error, so we can replace `expect`'s below
     /// with proper error handling.
-    fn xform_collection<'a, 'b, P, T, LC>(
-        col: Collection<Child<'a, P, T>, DDValue, Weight>,
+    // TODO: Much of this logic would be vastly simplified if we used a
+    //       combination of traits and `Vec<XFormCollection>`s (as opposed to
+    //       what we do now with a linked list of them)
+    fn xform_collection<'a, S, T, Lookup>(
+        col: Collection<S, DDValue, Weight>,
         xform: &Option<XFormCollection>,
-        arrangements: &Arrangements<'a, 'b, P, T>,
-        lookup_collection: &LC,
+        arrangements: &Arrangements<'a, S, T>,
+        lookup_collection: Lookup,
         is_top_level_scope: bool,
```

Comment:
I think this flag should no longer be needed.

Reply:
For strongly connected rules we pass `false` to `is_top_level_collection`, I guess I'd make a `mk_rule_streamless` function for that? I really don't like it, but I'll try and address the code duplication in successive PRs

Reply:
Nah, let's not introduce code duplication in the first place. Just ignore my comment.

Reply:
Actually, it may be 100% fine to allow `mk_rule()` to create recursive streams, I'm pretty sure it's valid within timely

Reply:
PS. I am still reviewing this and might have more comments.
PPS. It looks like tests are now passing (yay!) except some linter errors.

Reply:
I've fixed the linting stuff locally, just wanted to push as many changes at once as possible to reduce CI times since I'm impatient